Fix Abs.n() not evaluating constants.

utam-1 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #25765

#### Brief description of what is fixed or changed
##### Modified sympy/core/function.py:
- Modified `_eval_evalf` in `class Function` to handle partial expressions.

##### Modified tests in following files:
- `sympy\codegen\tests\test_algorithms.py`: Loosened tolerance to account for changes. (Based on limits set for machine epsilon by IEEE- 754)
- `sympy\physics\control\tests\test_lti.py`: Changed test cases to expect evaluation.
- `sympy\simplify\tests\test_simplify.py`: Changed test case to expect floating-point conversion. 

#### Other comments
The mentioned issue wasn't strictly confined to `Abs`, but for every function that is associated with Function class in `function.py`

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* functions
  * Fixed .n() not evaluating expressions inside functions.

<!-- END RELEASE NOTES -->

utam-1 committed:
Fix Abs.n() not evaluating constants

Implemented _eval_evalf for Abs class to evaluate constants
while preserving symbols.

utam-1 committed:
Use doctest ellipsis

utam-1 committed:
Removed Examples.

utam-1 commented:
@oscarbenjamin, I was working on a fix for this issue. Would you please review this PR and let me know if this approach is correct?

oscarbenjamin commented:
This is not the right approach.

Fixing the issue requires taking the time to go through the whole evalf stack and think about how to restructure it. Adding more `_eval_evalf` methods in the wrong place that also call the wrong methods is just going to make everything more of a mess than it already is.

It looks to me like the code here is generated by an LLM. While LLMs can generate quick hacky code they are not a substitute for taking time to understand the codebase. I think if you study the whole mechanism of how evalf works then it should be quite obvious that we don't just want to add an `_eval_evalf` method to a random class.

Studying the whole of how evalf works is not easy but it is what is actually required to fix this issue. This is not something where asking an LLM to just write some code and tests is going to result in a reasonable fix.

utam-1 committed:
Revert "Removed Examples."

This reverts commit 35f829ca64587ca1b3487e7fa74b4e3a9e79079e.

smichr commented:
If we make the fallback in Function._eval_evalf be the following, then functions that have a known mpmath function but have symbols that prevent mpmath from evaluating can be evaluated:
```python
        # Fall-back evaluation
        if func is None:
            imp = getattr(self, '_imp_', None)
            if imp is None:
                return None
            try:
                return Float(imp(*[i.evalf(prec) for i in self.args]), prec)
            except (TypeError, ValueError):
                return None
        elif not all(_.is_number for _ in args):  # mpmath will not handle symbols
            args = [_.evalf(prec, maxn=1) for _ in args]
            return self.func(*args)
```
For example
```
>>> sin(exp(1)+1+x).n(3)
sin(x + 3.718281828459)
>>> abs(exp(1)+1+x).n(3)
Abs(x + 3.718281828459)
>>> f=Function('f')
>>> f(exp(1)+1+x).n(3)  # the handling in the mpmath loop would have to change for this but maybe it is better to leave as is
f(x + 1 + E)
```

utam-1 commented:
> If we make the fallback in Function._eval_evalf be the following, then functions that have a known mpmath function but have symbols that prevent mpmath from evaluating can be evaluated:
> 
> ```python
>         # Fall-back evaluation
>         if func is None:
>             imp = getattr(self, '_imp_', None)
>             if imp is None:
>                 return None
>             try:
>                 return Float(imp(*[i.evalf(prec) for i in self.args]), prec)
>             except (TypeError, ValueError):
>                 return None
>         elif not all(_.is_number for _ in args):  # mpmath will not handle symbols
>             args = [_.evalf(prec, maxn=1) for _ in args]
>             return self.func(*args)
> ```
> 
> For example
> 
> ```
> >>> sin(exp(1)+1+x).n(3)
> sin(x + 3.718281828459)
> >>> abs(exp(1)+1+x).n(3)
> Abs(x + 3.718281828459)
> >>> f=Function('f')
> >>> f(exp(1)+1+x).n(3)  # the handling in the mpmath loop would have to change for this but maybe it is better to leave as is
> f(x + 1 + E)
> ```

@smichr, thanks for this suggested fix. While investigating this issue I studied how Max and Min handle partial evaluation, since they work correctly with mixed symbolic/numeric expressions. I found that `_eval_evalf` in `class AssocOp(Basic)` handled mixed expressions similarly : – 
```
args = []
for a in self.args:
    newa = a._eval_evalf(prec)
    if newa is None:
        args.append(a)
    else:
        args.append(newa)
return self.func(*args)
```
Furthermore, I think leaving undefined function as it is plausible, because users might want to keep the current behavior of undefined functions intact for symbolic manipulations (possibly). If it is needed in future, we might be able to address it.

smichr commented:
That version for AssocOp is more defensive. Looks like a good model to follow.

utam-1 committed:
Added fallback for mixed expression.

Modified the fix code suggested to match similar implementation shown in AssocOp._eval_evalf.

utam-1 committed:
Improved exception handling.

utam-1 committed:
Update function.py

utam-1 committed:
Modified check for symbols.

not is_number check was broad and was catching Tuples/Lists as well.

utam-1 committed:
Updated code based on discussion.

Changes made : –
- Placed the fallback mechanism only after mpmath fails.

- Modified test cases to account for changes.

- Left trig functions as it is, as the users may want them unevaluated.

utam-1 commented on sympy/core/function.py:
```diff
@@ -561,7 +561,18 @@ def _get_mpmath_func(fname):
             except (TypeError, ValueError):
                 return None
-
+        elif any(arg.free_symbols for arg in args): # mpmath will not handle symbols
```
Yes, I believe that seems plausible. I have avoided trig functions altogether in my changes. Following-up to this, I would like to ask that wouldn't the same behavior be expected from other functions as well (except Abs possibly) under certain conditions?

utam-1 committed:
Removed constraint for trig functions.

Changed test cases and removed check for identifying trig functions.

utam-1 commented:
> But the request for evaluation _of current numbers_ in an expression, whether or not there are other symbols is a reasonable state to request. So perhaps all functions, even unevaluated, should simply do as much evaluation as possible. It would be nice to get some input from any others on this.

I agree with what you said, furthermore, based on the information I've gathered, I believe calling `.evalf()` for any function should be the last step whenever substitution (or any other evaluation is considered) before it. For example: – 
```
# Symbolic operations first, then evaluate
expr = sin(theta*t + pi/4)
shifted = expr.subs(theta*t + pi/4, theta*t)      # Symbolic manipulation
result = shifted.n()           # Evaluate at the end -----> Correct behavior, gives sin(theta*t). 

#  Evaluating too early
expr = sin(theta*t + pi/4).n()  # pi/4 → 0.785...
shifted = expr.subs(theta*t + pi/4, theta*t) --------> Subs didn't work as expected, gives sin(theta*t + 0.7853....)
``` 

```
# Perform substitution then evaluation
expr = sin(x/pi + sqrt(2)*t)
result = expr.subs(x, 2*pi).simplify()  # -> sin(√2⋅t + 2)
final = result.evalf()  # Evaluate at end -> sin(1.4142135623731⋅t + 2.0)

# Calling evalf early
early_eval = expr.evalf() # -> sin(1.4142135623731⋅t + 0.318309886183791⋅x)
result = early_eval.subs(x, 2*pi).simplify() # - > sin(1.4142135623731⋅t + 0.636619772367581⋅π)
```
This caution must be exercised by anyone who uses `evalf`.

utam-1 commented:
> Your comments indicate that something else happens but I don't see where that comes from unless it is an LLM hallucination.

I apologize that I didn't clarify it beforehand, but the code examples are meant to be run after the changes I proposed are applied.
```
from sympy import symbols, sin, pi, sqrt, pprint
theta = symbols('theta')
expr = sin(theta*t + pi/4).n() # Will give output as sin(t⋅θ + 0.785398163397448).
pprint(expr)

```
> trig functions should not be evaluated unless they are symbol free so that if symbols are replaced with pi-related values they might combine with pi multiples in the numerical portion of the args e.g. sin(x/pi) will be easy to evaluate at x = 2*pi.                       

I just wanted to give my views on the above comment by @smichr, i.e, even after the changes are applied caution must be exercised before using `evalf`, and it is always good to perform other evaluations before resorting to `evalf`.

oscarbenjamin commented:
I think this is making evalf into more of a mess by adding more code in the wrong place.

utam-1 commented:
> This is looking good. Perhaps you can add a test like f(x + pi).n(3) to show that the pi gets evaluated even in an undefined function.

@smichr, the implementation that I have given here is currently **not** supporting UDFs, however locally I have tried changing the following in function.py : – 

```
# Fall-back evaluation
        if func is None:
            imp = getattr(self, '_imp_', None)
            if imp is not None:
                try:
                    return Float(imp(*[i.evalf(prec) for i in self.args]), prec)
                except (TypeError, ValueError):
                    return None
            if all(isinstance(arg, Expr) for arg in self.args): ------> Basically do what we did for normal functions.
                dps = prec_to_dps(prec)
                new_args = []
                is_changed= False
                for arg in self.args:
                    try:
                        new_arg = arg.evalf(dps, maxn=1)
                        if new_arg is None or new_arg == arg: # could not evalf hence fallback to original
                            new_args.append(arg)
                        else:
                            new_args.append(new_arg)
                            is_changed = True ---------> Keep track of changes
                    except (TypeError, ValueError, AttributeError):
                        new_args.append(arg) # Use original args even in case of exception.
                if is_changed:
                    return self.func(*new_args, evaluate=False) -----> This was needed to prevent reconstruction after evaluation, which resulted in sort of Recursion Error.
                else:
                    return None

```
### This resulted in following test failures, most of these can be fixed if test cases are updated in my based on what I have understood, still it would be better if you gave more insight on this ->

```
def test_nfloat():
        from sympy.core.basic import _aresame
        from sympy.polys.rootoftools import rootof

        x = Symbol("x")
        eq = x**Rational(4, 3) + 4*x**(S.One/3)/3
        assert _aresame(nfloat(eq), x**Rational(4, 3) + (4.0/3)*x**(S.One/3))
        assert _aresame(nfloat(eq, exponent=True), x**(4.0/3) + (4.0/3)*x**(1.0/3))
        eq = x**Rational(4, 3) + 4*x**(x/3)/3
        assert _aresame(nfloat(eq), x**Rational(4, 3) + (4.0/3)*x**(x/3))
        big = 12345678901234567890
        # specify precision to match value used in nfloat
        Float_big = Float(big, 15)
        assert _aresame(nfloat(big), Float_big)
        assert _aresame(nfloat(big*x), Float_big*x)
        assert _aresame(nfloat(x**big, exponent=True), x**Float_big)
        assert nfloat(cos(x + sqrt(2))) == cos(x + nfloat(sqrt(2)))

        # issue 6342
        f = S('x*lamda + lamda**3*(x/2 + 1/2) + lamda**2 + 1/4')
        assert not any(a.free_symbols for a in solveset(f.subs(x, -0.139)))

        # issue 6632
        assert nfloat(-100000*sqrt(2500000001) + 5000000001) == \
            9.99999999800000e-11

        # issue 7122
        eq = cos(3*x**4 + y)*rootof(x**5 + 3*x**3 + 1, 0)
        assert str(nfloat(eq, exponent=False, n=1)) == '-0.7*cos(3.0*x**4 + y)'

        # issue 10933
        for ti in (dict, Dict):
            d = ti({S.Half: S.Half})
            n = nfloat(d)
            assert isinstance(n, ti)
            assert _aresame(list(n.items()).pop(), (S.Half, Float(.5)))
        for ti in (dict, Dict):
            d = ti({S.Half: S.Half})
            n = nfloat(d, dkeys=True)
            assert isinstance(n, ti)
            assert _aresame(list(n.items()).pop(), (Float(.5), Float(.5)))
        d = [S.Half]
        n = nfloat(d)
        assert type(n) is list
        assert _aresame(n[0], Float(.5))
        assert _aresame(nfloat(Eq(x, S.Half)).rhs, Float(.5))
        assert _aresame(nfloat(S(True)), S(True))
        assert _aresame(nfloat(Tuple(S.Half))[0], Float(.5))
        assert nfloat(Eq((3 - I)**2/2 + I, 0)) == S.false
        # pass along kwargs
        assert nfloat([{S.Half: x}], dkeys=True) == [{Float(0.5): x}]

        # Issue 17706
        A = MutableMatrix([[1, 2], [3, 4]])
        B = MutableMatrix(
            [[Float('1.0', precision=53), Float('2.0', precision=53)],
            [Float('3.0', precision=53), Float('4.0', precision=53)]])
        assert _aresame(nfloat(A), B)
        A = ImmutableMatrix([[1, 2], [3, 4]])
        B = ImmutableMatrix(
            [[Float('1.0', precision=53), Float('2.0', precision=53)],
            [Float('3.0', precision=53), Float('4.0', precision=53)]])
        assert _aresame(nfloat(A), B)

        # issue 22524
        f = Function('f')
>       assert not nfloat(f(2)).atoms(Float)
E       assert not {2.00000000000000}
E        +  where {2.00000000000000} = atoms(Float)
E        +    where atoms = f(2.0).atoms
E        +      where f(2.0) = nfloat(f(2))
E        +        where f(2) = f(2)

sympy\core\tests\test_function.py:982: AssertionError
_____________________________________________________________ test_StateSpace_dsolve _____________________________________________________________
[gw0] win32 -- Python 3.11.4 C:\Users\uttam\AppData\Local\Programs\Python\Python311\python.exe

    def test_StateSpace_dsolve():
        # https://web.mit.edu/2.14/www/Handouts/StateSpaceResponse.pdf
        # https://lpsa.swarthmore.edu/Transient/TransMethSS.html
        A1 = Matrix([[0, 1], [-2, -3]])
        B1 = Matrix([[0], [1]])
        C1 = Matrix([[1, -1]])
        D1 = Matrix([0])
        I1 = Matrix([[1], [2]])
        t = symbols('t')
        ss1 = StateSpace(A1, B1, C1, D1)

        # Zero input and Zero initial conditions
        assert ss1.dsolve() == Matrix([[0]])
        assert ss1.dsolve(initial_conditions=I1) == Matrix([[8*exp(-t) - 9*exp(-2*t)]])

        A2 = Matrix([[-2, 0], [1, -1]])
        C2 = eye(2,2)
        I2 = Matrix([2, 3])
        ss2 = StateSpace(A=A2, C=C2)
        assert ss2.dsolve(initial_conditions=I2) == Matrix([[2*exp(-2*t)], [5*exp(-t) - 2*exp(-2*t)]])

        A3 = Matrix([[-1, 1], [-4, -4]])
        B3 = Matrix([[0], [4]])
        C3 = Matrix([[0, 1]])
        D3 = Matrix([0])
        U3 = Matrix([10])
        ss3 = StateSpace(A3, B3, C3, D3)
        op = ss3.dsolve(input_vector=U3, var=t)
        assert str(op.simplify().expand().evalf()[0]) == str(5.0 + 20.7880460155075*exp(-2.5*t)*sin(1.3228756555323*t)
                                                - 5.0*exp(-2.5*t)*cos(1.3228756555323*t))

        # Test with Heaviside as input
        A4 = Matrix([[-1, 1], [-4, -4]])
        B4 = Matrix([[0], [4]])
        C4 = Matrix([[0, 1]])
        U4 = Matrix([[10*Heaviside(t)]])
        ss4 = StateSpace(A4, B4, C4)
        op4 = str(ss4.dsolve(var=t, input_vector=U4)[0].simplify().expand().evalf())
>       assert op4 == str(5.0*Heaviside(t) + 20.7880460155075*exp(-2.5*t)*sin(1.3228756555323*t)*Heaviside(t)
                                                - 5.0*exp(-2.5*t)*cos(1.3228756555323*t)*Heaviside(t))
E       AssertionError: assert '5.0*Heavisid...iside(t, 0.5)' == '5.0*Heavisid...*Heaviside(t)'
E
E         - 5.0*Heaviside(t) + 20.7880460155075*exp(-2.5*t)*sin(1.3228756555323*t)*Heaviside(t) - 5.0*exp(-2.5*t)*cos(1.3228756555323*t)*Heaviside(t)
E         + 5.0*Heaviside(t, 0.5) + 20.7880460155075*exp(-2.5*t)*sin(1.3228756555323*t)*Heaviside(t, 0.5) - 5.0*exp(-2.5*t)*cos(1.3228756555323*t)*Heaviside(t, 0.5)
E         ?                +++++                                                                   +++++                                           
           +++++

sympy\physics\control\tests\test_lti.py:3578: AssertionError
________________________________________________________________ test_fcode_sign _________________________________________________________________
[gw0] win32 -- Python 3.11.4 C:\Users\uttam\AppData\Local\Programs\Python\Python311\python.exe

    def test_fcode_sign():  #issue 12267
        x=symbols('x')
        y=symbols('y', integer=True)
        z=symbols('z', complex=True)
        assert fcode(sign(x), standard=95, source_format='free') == "merge(0d0, dsign(1d0, x), x == 0d0)"
        assert fcode(sign(y), standard=95, source_format='free') == "merge(0, isign(1, y), y == 0)"
>       assert fcode(sign(z), standard=95, source_format='free') == "merge(cmplx(0d0, 0d0), z/abs(z), abs(z) == 0d0)"
E       AssertionError: assert 'merge(cmplx(...bs(z) == 0d0)' == 'merge(cmplx(...bs(z) == 0d0)'
E
E         - merge(cmplx(0d0, 0d0), z/abs(z), abs(z) == 0d0)
E         ?              --  --
E         + merge(cmplx(0, 0), z/abs(z), abs(z) == 0d0)

sympy\printing\tests\test_fortran.py:62: AssertionError
________________________________________________________________ test_issue_22523 ________________________________________________________________
[gw1] win32 -- Python 3.11.4 C:\Users\uttam\AppData\Local\Programs\Python\Python311\python.exe

    def test_issue_22523():
        N, s = symbols('N s')
        rho = Function('rho')
        # intentionally use 4.0 to confirm issue with nfloat
        # works here
        eqn = 4.0*N*sqrt(N - 1)*rho(s) + (4*s**2*(N - 1) + (N - 2*s*(N - 1))**2
            )*Derivative(rho(s), (s, 2))
        match = classify_ode(eqn, dict=True, hint='all')
        assert match['2nd_power_series_ordinary']['terms'] == 5
        C1, C2 = symbols('C1,C2')
        sol = dsolve(eqn, hint='2nd_power_series_ordinary')
        # there is no r(2.0) in this result
>       assert filldedent(sol) == filldedent(str('''
            Eq(rho(s), C2*(1 - 4.0*s**4*sqrt(N - 1.0)/N + 0.666666666666667*s**4/N
            - 2.66666666666667*s**3*sqrt(N - 1.0)/N - 2.0*s**2*sqrt(N - 1.0)/N +
            9.33333333333333*s**4*sqrt(N - 1.0)/N**2 - 0.666666666666667*s**4/N**2
            + 2.66666666666667*s**3*sqrt(N - 1.0)/N**2 -
            5.33333333333333*s**4*sqrt(N - 1.0)/N**3) + C1*s*(1.0 -
            1.33333333333333*s**3*sqrt(N - 1.0)/N - 0.666666666666667*s**2*sqrt(N
            - 1.0)/N + 1.33333333333333*s**3*sqrt(N - 1.0)/N**2) + O(s**6))'''))
E       AssertionError: assert '\nEq(rho(s),...N) + O(s**6))' == '\nEq(rho(s),...2) + O(s**6))'
E
E
E         - Eq(rho(s), C2*(1 - 4.0*s**4*sqrt(N - 1.0)/N + 0.666666666666667*s**4/N
E         - - 2.66666666666667*s**3*sqrt(N - 1.0)/N - 2.0*s**2*sqrt(N - 1.0)/N +
E         + Eq(rho(s), -1.33333333333333*s**3*r(2.0)/N - 2.0*s**4*r(3.0)/N +
E         + 1.33333333333333*s**3*r(2.0) + 2.0*s**4*r(3.0) + C2*(1 +
E         - 9.33333333333333*s**4*sqrt(N - 1.0)/N**2 - 0.666666666666667*s**4/N**2...
E
E         ...Full output truncated (12 lines hidden), use '-vv' to show

sympy\solvers\ode\tests\test_ode.py:1061: AssertionError

```
